geopandas/tools/overlay.py

```diff
@@ -317,32 +317,61 @@ def _make_valid(df):
             result = dfunion[dfunion["__idx1"].notnull()].copy()
 
     if keep_geom_type:
-        key_order = result.keys()
-        exploded = result.reset_index(drop=True).explode()
-        exploded = exploded.reset_index(level=0)
+        geom_type = df1.geom_type.iloc[0]
+
+        # First we filter the geometry types inside GeometryCollections objects
+        # (e.g. GeometryCollection([polygon, point]) -> polygon)
+        # we do this separately on only the relevant rows, as this is an expensive
+        # operation (an expensive no-op for geometry types other than collections)
+        is_collection = result.geom_type == "GeometryCollection"
+        if is_collection.any():
+            geom_col = result._geometry_column_name
+            collections = result[[geom_col]][is_collection]
+
+            exploded = collections.reset_index(drop=True).explode()
+            exploded = exploded.reset_index(level=0)
+
+            orig_num_geoms_exploded = exploded.shape[0]
+            if geom_type in polys:
+                exploded = exploded.loc[exploded.geom_type.isin(polys)]
+            elif geom_type in lines:
+                exploded = exploded.loc[exploded.geom_type.isin(lines)]
+            elif geom_type in points:
+                exploded = exploded.loc[exploded.geom_type.isin(points)]
+            else:
+                raise TypeError(
+                    "`keep_geom_type` does not support {}.".format(geom_type)
+                )
+            num_dropped_collection = orig_num_geoms_exploded - exploded.shape[0]
 
+            # level_0 created with above reset_index operation
+            # and represents the original geometry collections
+            dissolved = exploded.dissolve(by="level_0")
```

Comment:
Another space for further speed-up is to avoid `dissolve` calling `unary_union` as we know that we just want to combine geometries in their respective Multi* versions. So we could have here something like `collect` (I am not sure if the existing function actually works) https://github.com/geopandas/geopandas/blob/5c0c9da1d7b13a78f4ffdd2b55ce61f88b60fd3d/geopandas/tools/util.py#L13

Happy to keep it as a follow-up (can you make a TODO note in the comment?

Reply:
Added a comment